README.md
# ExposureResearchTools
Exposure Research Tools (Matomo Plugin)

This plugin restructures data collected via Matomo (formerly known as Piwik) to be used in Selective Exposure research.
It extracts an iditification variable from the first visit's GET request to allow merging survey data with observational data from Matomo.

The tools adds a menu item "Research Tools" -> "Export Visits" to the Matomo menu.

## Background details

For a detailed discussion of the tool and its application in selective exposure research, please refer to our paper.

Leiner, D. J., Scherr, S., & Bartsch, A. (2016). Using Open-Source Tools to Measure Online Selective Exposure in Naturalistic Settings. Communication Methods and Measures, 10(4), 199–216. [doi:10.1080/19312458.2016.1224825](http://doi.org/10.1080/19312458.2016.1224825)

## Resources

We have prepared a template for [SoSci Survey](http://www.soscisurvey.com). The template ...

* opens a pop-up window with the stimulus (that should be monitored by Matomo to collect selective exposure data)
* stores the times when the pop-up was opened and closed (to compute the browsing time)
* closes the pop-up after a predefined time (optionally)
* supports pre and post surveys

Using the template:

* [Download the template](https://raw.githubusercontent.com/BurninLeo/ExposureResearchTools/master/resources/sosci.template.xml) -> Right-click + Save as
* Register an account on [SoSci Survey](http://www.soscisurvey.com) and create a new survey project
* Go to **Survey Project** -> **Project Settings** -> **Import Project or Questions**
* Select the XML file you just downloaded and import
* Go to **Compose Questionnaire** and try with the green play button

Opening a pop-up in a different survey software:

* There is a [JavaScript Example](https://github.com/BurninLeo/ExposureResearchTools/blob/master/resources/JavaScript%20Sample.txt) that could be used on other software that SoSci Survey.

**Note:** To collect selective exposure data, yo need a Matomo installation and must edit the stimulus website to report to Matomo. Please read the instructions in the paper (Leiner et al., forthcoming) for details.


CHANGELOG.md
# Changelog

* 0.1.0 Original release of the plugin (stable, 2016-07-04)
* 0.1.4 Added resources (stable, 2016-10-06)
  * Resources (JavaScript example and SoSci Survey template) were added to integrate a selective exposure experiment into a survey
  * A reference to the resources was added in the README, in the plugin description, and in the plugin itself
* 0.1.7 Compatibility with Piwik 3.x
  * Changes report to appear in Piwik 3.x
  * Shorten variable names for page times to comply with SPSS (max. 64 characters)
* 0.1.8 Cleaning
  * Removed two undefined variable issues that triggered error log entries

API.php
<?php
/**
 * Piwik - free/libre analytics platform
 *
 * @link http://piwik.org
 * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
 *
 */
namespace Piwik\Plugins\ExposureResearchTools;

use Piwik\DataTable;
use Piwik\DataTable\Row;

/**
 * API for plugin ExposureResearchTools
 *
 * @method static \Piwik\Plugins\ExposureResearchTools\API getInstance()
 */
class API extends \Piwik\Plugin\API
{
	
	
	/**
	 * From: https://en.wikibooks.org/wiki/Algorithm_Implementation/Strings/Longest_common_substring#PHP
	 * @param string $string_1
	 * @param string $string_2
	 * @return string
	 */
	private static function get_longest_common_subsequence($string_1, $string_2)
	{
		$string_1_length = strlen($string_1);
		$string_2_length = strlen($string_2);
		$return          = '';
	
		if ($string_1_length === 0 || $string_2_length === 0)
		{
			// No similarities
			return $return;
		}
	
		$longest_common_subsequence = array();
	
		// Initialize the CSL array to assume there are no similarities
		$longest_common_subsequence = array_fill(0, $string_1_length, array_fill(0, $string_2_length, 0));
	
		$largest_size = 0;
	
		for ($i = 0; $i < $string_1_length; $i++)
		{
			for ($j = 0; $j < $string_2_length; $j++)
			{
				// Check every combination of characters
				if ($string_1[$i] === $string_2[$j])
				{
					// These are the same in both strings
					if ($i === 0 || $j === 0)
					{
						// It's the first character, so it's clearly only 1 character long
						$longest_common_subsequence[$i][$j] = 1;
					}
					else
					{
						// It's one character longer than the string from the previous character
						$longest_common_subsequence[$i][$j] = $longest_common_subsequence[$i - 1][$j - 1] + 1;
					}
	
					if ($longest_common_subsequence[$i][$j] > $largest_size)
					{
						// Remember this as the largest
						$largest_size = $longest_common_subsequence[$i][$j];
						// Wipe any previous results
						$return       = '';
						// And then fall through to remember this new value
					}
	
					if ($longest_common_subsequence[$i][$j] === $largest_size)
					{
						// Remember the largest string(s)
						$return = substr($string_1, $i - $largest_size + 1, $largest_size);
					}
				}
				// Else, $CSL should be set to 0, which it was already initialized to
			}
		}
	
		// Return the list of matches
		return $return;
	}

    /**
     * Another example method that returns a data table.
     * @param int    $idSite
     * @param string $period
     * @param string $date
     * @param bool|string $segment
     * @return DataTable
     */
    public function getData($idSite, $period, $date, $segment=false)
    {
		// Not: \Piwik is \core
		// TODO: Zugriffssteuerung !!!
    	\Piwik\Piwik::checkUserHasViewAccess($idSite);
    	
    	// Read additional variables
    	$aLimit = (int)\Piwik\Common::getRequestVar('vlimit', 100, 'int');
    	if ($aLimit === 0) {
    		$aLimit = 100;
    	}
    	$getVar = trim((string)\Piwik\Common::getRequestVar('casevar', '', 'string'));
    	$clipServer = ((string)\Piwik\Common::getRequestVar('server', 'no', 'string') !== 'yes');
    	$clipExtension = ((string)\Piwik\Common::getRequestVar('noclip', 'no', 'string') !== 'yes');
    	$recodeIndex = ((string)\Piwik\Common::getRequestVar('disindex', 'no', 'string') !== 'yes');
    	$skipNoID = ((string)\Piwik\Common::getRequestVar('skipid', 'no', 'string') === 'yes');
    	$addAggregate = ((string)\Piwik\Common::getRequestVar('aggregate', 'no', 'string') === 'yes');
    	$structure = (string)\Piwik\Common::getRequestVar('structure', 'case', 'string');  // [case|page]
    	$var64 = ((string)\Piwik\Common::getRequestVar('var64', 'no', 'string') === 'yes');

		// https://developer.piwik.org/guides/security-in-piwik
		// https://developer.piwik.org/guides/persistence-and-the-mysql-backend

		$sql =
			'SELECT v.idvisit, l.server_time, l.time_spent_ref_action, a.name, a.idaction
			FROM '.\Piwik\Common::prefixTable('log_visit').' v
			LEFT JOIN '.\Piwik\Common::prefixTable('log_link_visit_action').' l ON (v.idvisit = l.idvisit)
			LEFT JOIN '.\Piwik\Common::prefixTable('log_action').' a ON (a.idaction = l.idaction_url)
			WHERE (v.idsite = ?) AND (a.type = 1)
			ORDER BY l.idvisit, l.server_time';
		
		$res = \Piwik\Db::query($sql, array($idSite));
		
		// error_log(var_export($idSite, true));
		
		if (empty($getVar)) {
			$getVarLen = 0;
			$skipNoID = false;
		} else {
			$getVarLen = strlen($getVar) + 1;
		}
		
		$nAction = 0;
		$nActionMax = 0; // Number of actions per visit (max.)
    	$cvVisitID = null;
    	$cvCaseID = null;
    	$cvCaseStart = null;
    	$cvTimeSum = 0;
    	$cvCaseActions = array();
    	$cvCaseTimes = array();
    	$cvCaseStamps = array();
    	$visits = array();
    	$actionIDs = array();  // Collect IDs per action
		do {
			$row = $res->fetch();
			
			// Stop processing for empty results
			if ($row === false) {
				// ... after the previous visit was registered
				if ($cvVisitID === null) {
					// No data at all
					$table->addRowFromSimpleArray(array('error' => 'no data available', 'site' => $idSite));
					return $table;
				}
			}
			$cvFirst = false;
			
			// Store new visit
			if (($row === false) or (($row['idvisit'] !== $cvVisitID) and ($cvVisitID !== null))) {
				if ($skipNoID and ($cvCaseID === null)) {
					// Skip this case
				} else {
					$visits[] = array(
						'id' => $cvVisitID,
						'case' => $cvCaseID,
						'start' => $cvCaseStart,
						'actions' => $cvCaseActions,
						'times' => $cvCaseTimes,
						'stamps' => $cvCaseStamps,
						'time sum' => $cvTimeSum
					);
					if (count($cvCaseActions) > $nAction) {
						$nAction = count($cvCaseActions);
					}
					if ($nAction > $nActionMax) {
						$nActionMax = $nAction;
					}
				}
			}
			// Stop, if there is no more data
			if ($row === false) {
				break;
			}
			
			// Register new visit (group of actions/page views)
			if ($row['idvisit'] !== $cvVisitID) {
				$nAction = 1;
				$cvVisitID = $row['idvisit'];
				$cvCaseStart = $row['server_time'];
				$cvTimeSum = 0;
				$cvCaseID = null;
				$cvCaseActions = array();
    			$cvCaseTimes = array();
    			$cvCaseStamps = array();
			} else {
				$nAction++;
			}
			
			// Decode the URL
			$url = $row['name'];
			if (($cvCaseID === null) and ($getVarLen !== 0)) {
				if (($p = strpos($url, '?')) !== false) {
					$qs = substr($url, $p + 1);
					$items = explode('&', $qs);
					foreach ($items as $item) {
						if (substr($item, 0, $getVarLen) === $getVar.'=') {
							$s = trim(substr($item, $getVarLen));
							if ($s !== '') {
								$cvCaseID = $s;
							}
						}
					}
				}
			}
			if ($recodeIndex) {
				$t10 = substr($url, -10);
				$t11 = substr($url, -11);
				if (($t10 === '/index.php') or ($t10 === '/index.htm') or ($t11 === '/index.html')) {
					// No index in subdirectories
					if (substr_count($url, '/') === 1) {
						$p = strpos($url, '/');
						$url = substr($url, 0, $p+1);
					}
				}
			}
			if ($clipServer) {
				$p = strpos($url, '/');
				if ($p !== false) {
					$url = substr($url, $p+1);
					if ($url === false) {
						$url = '';
					}
				}
			}
			if ($clipExtension) {
				$p1 = strrpos($url, '.');
				$p2 = strrpos($url, '?');
				if (($p1 === false) and ($p2 === false)) {
					// No extension
					$p = false;
				} elseif ($p1 === false) {
					$p = $p2;
				} elseif ($p2 === false) {
					$p = $p1;
				} else {
					$p = min($p1, $p2);
				}
				if ($p !== false) {
					$url = substr($url, 0, $p);
				}
			}
			if ($url === '') {
				$url = '/';
			}
			
			$cvCaseActions[] = $url;
			$aTime = (int)$row['time_spent_ref_action'];
			$cvCaseTimes[] = $aTime;
			$cvCaseStamps[] = $row['server_time'];
			$cvTimeSum+= $aTime;
			
			// Already know this action?
			if (!isset($actionIDs[$url])) {
				$actionIDs[$url] = (int)$row['idaction'];
			}
		} while ($row !== false);
		
		// Variable names for actions
		$actionNames = array();
		foreach ($visits as $visit) {
			foreach ($visit['actions'] as $i => $action) {
				if (!isset($actionNames[$action])) {
					if ($action === '/') {
						$actionNames[$action] = 'index';
					} else {
						// Some basic cleaning
						$actionNames[$action] = preg_replace(array(
								'/[^a-z0-9]+/i',
								'/^_+/',
								'/_+$/',
								'/__+/'
							), array(
								'_',
								'',
								'',
								'_'
							),
							$action
						);
					}
				}
			}
		}
		// Sort by original URL
		ksort($actionNames);
		// Rename, if applicable
		if ($var64) {
			// Find common parts in all actions (except index) and remove
			$first = array();
			$common = '';
			foreach ($actionNames as $id => $name) {
				if ($id !== 'index') {
					$first[] = $name;
				}
				if (count($first) === 2) {
					$common = self::get_longest_common_subsequence($first[0], $first[1]);
					break;
				}
			}
			if (strlen($common) >= 3) {
				foreach ($actionNames as $id => $name) {
					if ($id === 'index') {
						continue;
					}
					$common = self::get_longest_common_subsequence($common, $name);
					if (strlen($common) < 3) {
						break;
					}
				}
			}
			// Remove common part
			if (strlen($common) >= 3) {
				foreach ($actionNames as $id => $name) {
					$actionNames[$id] = str_replace($common, '', $name);
				}
			}
			// Then also clip to 60 characters
			// (makes 63 with prefix AT_
			foreach ($actionNames as $id => $name) {
				if (strlen($name) > 60) {
					// TODO: Some time find a better algorith, using word boundaries (_) and skipping word in the middle
					$actionNames[$id] = substr($name, -60);
				}
			}
		}

		// Result table
		$table = new DataTable();
			
		if ($structure === 'page') {
			foreach ($visits as $visit) {
				$basic = array(
					'id' => $visit['id'],
					'CASE' => $visit['case']
				);
				$t0 = strtotime($visit['start']);
				
				// One row per action
				$aData = array();
				foreach ($visit['actions'] as $i => $action) {
					$aData['pos'] = ($i + 1);
					$aData['aID'] = (isset($actionIDs[$action]) ? $actionIDs[$action] : -1);
					$aData['url'] = $action;
					if (isset($visit['times'][$i+1])) {
						$aData['time'] = $visit['times'][$i+1];
					} else {
						$aData['time'] = null;
					}
					$aData['ontime'] = strtotime($visit['stamps'][$i]) - $t0;
					$aData['astime'] = $visit['stamps'][$i];
						
					$table->addRowFromSimpleArray(array_merge(
						$basic,
						$aData
					));
				}
			}
			
		} else {
			// Number of actions
			if ($nActionMax > $aLimit) {
				$nActionMax = $aLimit;
			}
			
			// Variable names for the page-time-variables
			if ($addAggregate) {
				$aggVarNames = [];
				foreach ($actionNames as $url => $actionID) {
					$aggVarNames[$url] = 'AT_'.$actionID;
				}
			}

			// Sort and package into table
			foreach ($visits as $visit) {
				$basic = array(
					'id' => $visit['id'],
					'CASE' => $visit['case'],
					'T0' =>  $visit['start']
				);
					
				// Rearrange data, if necessary
				$actions = [];
				$timesT = [];
				$timesAT = [];
				$pages = [];
				for ($i=0; $i<$nActionMax; $i++) {
					if (isset($visit['actions'][$i])) {
						$actions['A'.($i+1)] = $visit['actions'][$i];
					} else {
						$actions['A'.($i+1)] = null;
					}
					if (isset($visit['times'][$i+1])) {
						$timesT['T'.($i+1)] = $visit['times'][$i+1];
					} else {
						$timesT['T'.($i+1)] = null;
					}
					
				}
				
				if ($addAggregate) {
					foreach (array_keys($aggVarNames) as $url) {
						$timesAT[$aggVarNames[$url]] = null;
					}
					foreach ($visit['actions'] as $i => $url) {
						if ($timesAT[$aggVarNames[$url]] === null) {
							$timesAT[$aggVarNames[$url]] = (isset($visit['times'][$i+1]) ? $visit['times'][$i+1] : 0);
						} elseif (isset($visit['times'][$i+1])) {
							$timesAT[$aggVarNames[$url]]+= $visit['times'][$i+1];
						}
					}
				}
				
				$table->addRowFromSimpleArray(array_merge(
					$basic,
					$actions,
					$timesT,
					array('AT' => $visit['time sum']),
					$timesAT,
					$pages
				));
			}
		}
	        
        return $table;
    }
}


Reports/GetExposureResearchTools.php
<?php
/**
 * Piwik - free/libre analytics platform
 *
 * @link http://piwik.org
 * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
 */

namespace Piwik\Plugins\ExposureResearchTools\Reports;

use Piwik\Piwik;
use Piwik\Plugin\Report;
use Piwik\Plugin\ViewDataTable;
use Piwik\Plugins\Actions\Columns\ExitPageUrl;
use Piwik\View;

/**
 * This class defines a new report.
 *
 * See {@link http://developer.piwik.org/api-reference/Piwik/Plugin/Report} for more information.
 */
class GetExposureResearchTools extends Base
{
    protected function init()
    {
        parent::init();
        
        // Piwik 2.x
        $this->category = 'Research Tools';
        $this->name          = 'Export Visits'; // Piwik::translate('ExposureResearchToolsName');
        
        // Piwik 3.x
        $this->categoryId = 'Research Tools';
        $this->subcategoryId = 'Export Visits';
        
        $this->dimension     = new ExitPageUrl();
        $this->documentation = Piwik::translate('ExposureResearchToolsDocumentation');

        // This defines in which order your report appears in the mobile app, in the menu and in the list of widgets
        $this->order = 999;

        // By default standard metrics are defined but you can customize them by defining an array of metric names
        // $this->metrics       = array('nb_visits', 'nb_hits');

        // Uncomment the next line if your report does not contain any processed metrics, otherwise default
        // processed metrics will be assigned
        // $this->processedMetrics = array();

        // Uncomment the next line if your report defines goal metrics
        // $this->hasGoalMetrics = true;

        // Uncomment the next line if your report should be able to load subtables. You can define any action here
        // $this->actionToLoadSubTables = $this->action;

        // Uncomment the next line if your report always returns a constant count of rows, for instance always
        // 24 rows for 1-24hours
        // $this->constantRowsCount = true;

        // If a menu title is specified, the report will be displayed in the menu
        $this->menuTitle = $this->name;

        // If a widget title is specified, the report will be displayed in the list of widgets and the report can be
        // exported as a widget
        // $this->widgetTitle  = '';
    }

    /**
     * Here you can configure how your report should be displayed. For instance whether your report supports a search
     * etc. You can also change the default request config. For instance change how many rows are displayed by default.
     *
     * @param ViewDataTable $view
     */
    public function configureView(ViewDataTable $view)
    {
        if (!empty($this->dimension)) {
            $view->config->addTranslations(array('label' => $this->dimension->getName()));
        }

        // $view->config->show_search = false;
        // $view->requestConfig->filter_sort_column = 'nb_visits';
        // $view->requestConfig->filter_limit = 10';

        $view->config->columns_to_display = array_merge(array('label'), $this->metrics);
    }

    /**
     * Here you can define related reports that will be shown below the reports. Just return an array of related
     * report instances if there are any.
     *
     * @return \Piwik\Plugin\Report[]
     */
    public function getRelatedReports()
    {
        return array(); // eg return array(new XyzReport());
    }
    
    /**
     * A report is usually completely automatically rendered for you but you can render the report completely
     * customized if you wish. Just overwrite the method and make sure to return a string containing the content of the
     * report. Don't forget to create the defined twig template within the templates folder of your plugin in order to
     * make it work. Usually you should NOT have to overwrite this render method.
     *
     * @return string
     */
    public function render()
    {
    	/*
        $view = new View('@ExposureResearchTools/GetExposureResearchTools');
        $view->myData = array();
        return $view->render();
        */
        
    	$aLimit = (int)\Piwik\Common::getRequestVar('vlimit', 100, 'int');
    	if ($aLimit === 0) {
    		$aLimit = 100;
    	}
    	$getVar = (string)\Piwik\Common::getRequestVar('casevar', 'num', 'string');
        
        /*
        if ($send === 'yes') {
        	$api = \Piwik\Plugins\ExposureResearchTools\API::getInstance();
	        $data = $api->getExposureResearchTools(
	        	\Piwik\Common::getRequestVar('idSite'),
			    \Piwik\Common::getRequestVar('date'),
			    \Piwik\Common::getRequestVar('period'),
			    \Piwik\Common::getRequestVar('segment')
	        );
	        
	        var_dump($data);
	        
	        exit;
        } else {
        */
        
        	// Copy all parameters
        	/*
        	$params = '';
        	foreach ($_REQUEST as $key => $value) {
        		if (in_array($key, array('vlimit', 'casevar', 'format'))) {
        			continue;
        		}
        		$monkey = "\r\n".'<input type="hidden" name="'.htmlspecialchars($key).'" value="'.htmlspecialchars($value).'" />';
        	}
        	*/
        	
	        return
				'<form action="index.php" method="GET">
				<div>
					<input type="hidden" name="module" value="API" />
					<input type="hidden" name="method" value="ExposureResearchTools.getData" />
					<input type="hidden" name="idSite" value="'.\Piwik\Common::getRequestVar('idSite').'" />
					<input type="hidden" name="date" value="'.\Piwik\Common::getRequestVar('date').'" />
					<input type="hidden" name="period" value="'.\Piwik\Common::getRequestVar('period').'" />
					<!-- <input type="hidden" name="segment" value="'.\Piwik\Common::getRequestVar('segment', '').'" /> -->
					<input type="hidden" name="token_auth" value="'.\Piwik\Piwik::getCurrentUserTokenAuth().'" />
					<input type="hidden" name="format" value="csv" />
					<input type="hidden" name="filter_limit" value="-1" />
				</div>
				<table cellspacing="0" cellpadding="0">
				<colgroup>
					<col width="320">
					</col>
				</colgroup>
				<tr>
					<td style="padding-right: 1em">Limit activities per visit:</td>
					<td><input type="text" name="vlimit" value="'.$aLimit.'" style="width: 64px" /></td>
				</tr><tr>
					<td style="padding-right: 1em">Read subject ID from GET variable:</td>
					<td><input type="text" name="casevar" value="'.htmlspecialchars($getVar).'" style="width: 180px" /></td>
				</tr><tr>
					<td style="padding-right: 1em">Data structure:</td>
					<td>
						<select name="structure" size="1" style="max-width: 100%">
							<option value="case">Per participant (many variables, few rows)</option>
							<option value="page">Per page view (few variables, may rows)</option>
						</select>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<input type="checkbox" id="aggregate" name="aggregate" value="yes" checked="checked" />
						<label for="aggregate">
							Include aggregate reading times per page (sum per page)<br>
							<span style="font-size: 85%">(applicable in per-participant structure, only)</span>
						</label>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<input type="checkbox" id="var64" name="var64" value="yes" checked="checked" />
						<label for="var64">SPSS-compatible variable names</label>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<input type="checkbox" id="server" name="server" value="yes" />
						<label for="server">Retain domain name in URLs</label>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<input type="checkbox" id="noclip" name="noclip" value="yes" />
						<label for="noclip">Retain extension and query string in URLs (anything after the file name)</label>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<input type="checkbox" id="disindex" name="disindex" value="yes" />
						<label for="disindex">Distinguish index.html, index.htm, index.php, and homepage (/)</label>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<input type="checkbox" id="skipid" name="skipid" value="yes" />
						<label for="skipid">Skip visits without subject ID (only applicable if GET variable for subject ID is set)</label>
					</td>
				</tr><tr>
					<td colspan="2" style="padding-top: 0.5em; padding-bottom: 0.5em">
						<strong>Note:</strong> All available data will be exported, regardless of the period defined above.
					</td>
				</tr>
				</table>
							
				<div style="margin-top: 1em">
					<button type="submit">Download CSV</button>
				</div>
				</form>

				<h2>File Structure</h2>
				<p>Depending on the file structure selected above, the variables in the result will be:</p>
				<ul style="list-style: circle; padding-left: 24px; margin-bottom: 30px">
					<li>Case identification (both structures)
						<ul style="list-style: circle; padding-left: 24px">
							<li><strong style="width: 60px; display: inline-block">id</strong> Matomo\'s ID for the visit</li>
							<li><strong style="width: 60px; display: inline-block">CASE</strong> Case ID retrieved from the URL (see setting GET variable)</li>
						</ul>
					</li>
					<li>Structured &quot;per participant&quot;
						<ul style="list-style: circle; padding-left: 24px">
							<li><strong style="width: 60px; display: inline-block">T0</strong> Date and time when the first page was retrieved</li>
							<li><span style="width: 60px; display: inline-block"><strong>A1</strong>&ndash;<strong>A<i>n</i></strong></span> Activities performed (pages viewed) during the visit: ID of the first viewed page stored in A1, etc.</li>
							<li><span style="width: 60px; display: inline-block"><strong>T1</strong>&ndash;<strong>T<i>n</i></strong></span> Times spent per activity (time between retrieving one page and the following one): T1 is the time for A1 in seconds, etc.</li>
							<li><span style="width: 60px; display: inline-block"><strong>AT</strong></span> Times spent during the visit overall.</li>
							<li><span style="width: 60px; display: inline-block"><strong>AT_<i>xyz</i></strong></span> Aggregate time spent per activity <i>xyz</i> (e.g., page <i>xyz</i>), in seconds. This block of variables will only be available, if &quot;Include aggregate reading times&quot; has been checked above.</li>
						</ul>
					</li>
					<li>Structured &quot;per action&quot;
						<ul style="list-style: circle; padding-left: 24px">
							<li><span style="width: 60px; display: inline-block"><strong>pos</strong></span> Order of actions (page views) during the visit</li>
							<li><span style="width: 60px; display: inline-block"><strong>aID</strong></span> Unique ID for the action (page)</li>
							<li><span style="width: 60px; display: inline-block"><strong>url</strong></span> Description of the action (page URL, usually shortend)</li>
							<li><span style="width: 60px; display: inline-block"><strong>time</strong></span> Time spent on the action (page)</li>
							<li><span style="width: 60px; display: inline-block"><strong>ontime</strong></span> Beginning of action [sec], relative to the visit\'s first page retrieval</li>
							<li><span style="width: 60px; display: inline-block"><strong>astime</strong></span> Absolute timestamp of the action (page view)</li>
						</ul>
					</li>
				</ul>
				<p><strong>Note:</strong> Matomo will not record the time spent on the most recent page (activity). This affects <strong>T<i>n</i></strong>, <strong>AT</strong>, <strong>AT_<i>xyz</i></strong>, and <strong>time</strong>.</p>
				<p><strong>Note:</strong> If Excel won\'t open the file correctly (all data in one cell), download the CSV file to disk,
					then start Excel and open via menu &rarr; file &rarr; open. OpenOffice Calc will cause less trouble.</p>
							
				<h2>Additional Information</h2>
				<p>If you\'re interested in doing selective exposure research with this tool, the following may be helpful:</p>
				<ul style="list-style: circle; padding-left: 24px; margin-bottom: 30px">
					<li>
						<p>We published a paper about this tool and how to collect SE data with Matomo:</p>
						<p>Leiner, D. J., Scherr, S., & Bartsch, A. (2016). Using Open-Source Tools to Measure Online Selective Exposure in Naturalistic Settings. Communication Methods and Measures, 10(4), 199–216. doi:<a href="http://doi.org/10.1080/19312458.2016.1224825">10.1080/19312458.2016.1224825</a></p>
					</li>
					<li>
						<p>There are additional resources on the <a href="https://github.com/BurninLeo/ExposureResearchTools" target="_blank">Plugin Website</a> (on GitHub).</p>
						<p>Specifically, there\'s a template for SoSci Survey to embed a stimulus presentation (via pop-up) between a pre and post questionnaire.
							The template takes care of transmitting the respondent ID, so that it appears in the SE data,
							and to store the times when the pop-up was opened and closed.</p>
					</li>
				</ul>';
    }

    /**
     * By default your report is available to all users having at least view access. If you do not want this, you can
     * limit the audience by overwriting this method.
     *
     * @return bool
    public function isEnabled()
    {
        return Piwik::hasUserSuperUserAccess()
    }
     */
}